Tests/GRPCTests/MutualTLSTests.swift

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import EchoImplementation
+import EchoModel
+@testable import GRPC
+import GRPCSampleData
+import Logging
+import NIOConcurrencyHelpers
+import NIOCore
+import NIOPosix
+import NIOSSL
+import XCTest
+
+class MutualTLSTests: GRPCTestCase {
+  enum ExpectedOutcome {
+    case success
+    case serverError
+    case clientError
+  }
+
+  func performTestWith(
+    _ serverTLSConfiguration: GRPCTLSConfiguration?,
+    _ clientTLSConfiguration: GRPCTLSConfiguration?,
+    expect expectedOutcome: ExpectedOutcome
+  ) throws {
+    // Setup the server.
+    let serverEventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    var serverConfiguration = Server.Configuration.default(
+      target: .hostAndPort("localhost", 0),
+      eventLoopGroup: serverEventLoopGroup,
+      serviceProviders: [EchoProvider()]
+    )
+    serverConfiguration.tlsConfiguration = serverTLSConfiguration
+    let serverErrorExpectation = self.expectation(description: "server error")
+    serverErrorExpectation.isInverted = expectedOutcome != .serverError
+    serverErrorExpectation.assertForOverFulfill = false
+    let serverErrorDelegate = ServerErrorRecordingDelegate(expectation: serverErrorExpectation)
+    serverConfiguration.errorDelegate = serverErrorDelegate
+
+    let server = try! Server.start(configuration: serverConfiguration).wait()
+    let port = server.channel.localAddress!.port!
+
+    // Setup the client.
+    let clientEventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    var clientConfiguration = ClientConnection.Configuration.default(
+      target: .hostAndPort("localhost", port),
+      eventLoopGroup: clientEventLoopGroup
+    )
+    clientConfiguration.tlsConfiguration = clientTLSConfiguration
+    clientConfiguration.connectionBackoff = nil
+    clientConfiguration.backgroundActivityLogger = self.clientLogger
+    let clientErrorExpectation = self.expectation(description: "client error")
+    clientErrorExpectation.isInverted = expectedOutcome == .success
+    clientErrorExpectation.assertForOverFulfill = false
+    let clientErrorDelegate = ErrorRecordingDelegate(expectation: clientErrorExpectation)
+    clientConfiguration.errorDelegate = clientErrorDelegate
```

Comment:
Rather than using an error delegate, if you switch the the channel pool you will be able to retrieve locally produced error from the status (more new API!):

```swift
let call = client.get(...)
let status = try call.status.wait()
if let cause = status.cause {
  // ...
} else {
  // ...
}
```

Reply:
So I have some _interesting behaviour_ when I try and use the new pool API including tests blocking and also the error delegate just not working. I'm sure I'm probably holding it wrong.

In the meantime I think the PR is in a good shape o/w (pending what CI has to say, ofc).

Reply:
When an RPC requests a stream from the pool it will by default wait at most 30 seconds for a stream to become available, I suspect you're running into that rather than blocking indefinitely. FWIW I'm okay with sticking to client connection.